internal/controller/apiextensions/claim/syncer_csa.go

```diff
@@ -0,0 +1,213 @@
+	})
+
+	// If the bound XR already exists we want to restore its original external
+	// name annotation in order to ensure we don't try to rename anything after
+	// the fact.
+	if meta.WasCreated(xr) && en != "" {
+		meta.SetExternalName(xr, en)
+	}
+
+	// We want to propagate the claim's spec to the composite's spec, but first
+	// we must filter out any well-known fields that are unique to claims. We do
+	// this by:
+	// 1. Grabbing a map whose keys represent all well-known claim fields.
+	// 2. Deleting any well-known fields that we want to propagate.
+	// 3. Using the resulting map keys to filter the claim's spec.
+	wellKnownClaimFields := xcrd.CompositeResourceClaimSpecProps()
+	for _, field := range xcrd.PropagateSpecProps {
+		delete(wellKnownClaimFields, field)
+	}
+
+	// CompositionRevisionRef is a special field which needs to be propagated
+	// based on the Update policy. If the policy is `Manual`, we need to remove
+	// CompositionRevisionRef from wellKnownClaimFields, so it is propagated
+	// from the claim to the XR.
+	if xr.GetCompositionUpdatePolicy() != nil && *xr.GetCompositionUpdatePolicy() == xpv1.UpdateManual {
+		delete(wellKnownClaimFields, xcrd.CompositionRevisionRef)
+	}
+
+	cmSpec, ok := cm.Object["spec"].(map[string]any)
+	if !ok {
+		return errors.New(errUnsupportedClaimSpec)
+	}
+
+	// Propagate the claim's spec (minus well known fields) to the XR's spec.
+	xr.Object["spec"] = withoutKeys(cmSpec, xcrd.GetPropFields(wellKnownClaimFields)...)
+
+	// We overwrite the entire XR spec above, so we wait until this point to set
+	// the claim reference.
+	xr.SetClaimReference(cm.GetReference())
+
+	// If the claim references an XR, make sure we're going to apply that XR. We
+	// do this just in case the XR exists, but we couldn't get it due to a stale
+	// cache.
+	if ref := cm.GetResourceReference(); ref != nil {
+		xr.SetName(ref.Name)
+	}
+
+	// If the XR doesn't exist, derive a name from the claim. The generated name
+	// is likely (but not guaranteed) to be available when we create the XR. If
+	// taken, then we are going to update an existing XR, probably hijacking it
+	// from another claim.
+	if !meta.WasCreated(xr) {
+		xr.SetGenerateName(fmt.Sprintf("%s-", cm.GetName()))
+
+		// GenerateName is a no-op if the xr already has a name set.
+		if err := s.names.GenerateName(ctx, xr); err != nil {
+			return errors.Wrap(err, errGenerateName)
+		}
+	}
+
+	// We're now done syncing the XR from the claim. If this is a new XR it's
+	// important that we update the claim to reference it before we create it.
+	// This ensures we don't leak an XR. We could leak an XR if we created an XR
+	// then crashed before saving a reference to it. We'd create another XR on
+	// the next reconcile.
+	cm.SetResourceReference(meta.ReferenceTo(xr, xr.GetObjectKind().GroupVersionKind()))
+	if err := s.client.Update(ctx, cm); err != nil {
+		return errors.Wrap(err, errUpdateClaim)
+	}
+
+	// Apply the XR, unless it's a no-op change.
```

Comment:
```suggestion
	// Apply the XR.
```

nit: as per [the above](https://github.com/crossplane/crossplane/pull/5313/files#r1477655885) comment.